src/nn/graph.ts

```diff
@@ -151,84 +127,95 @@ export class MLGraph {
-    const outputOperands: Map<string, OutputOperand> = new Map();
-    for (const outputName in outputs) {
-      utils.assert(
-          typeof outputName === 'string' && this.outputs_.has(outputName),
-          'The name of the output is invalid.');
-      utils.assert(
-          utils.isTypedArray(outputs[outputName]),
-          'Only output of ArrayBufferView type is supported.');
-      outputOperands.set(outputName, this.outputs_.get(outputName));
+  private computeOutputTensors(
+    inputs: MLNamedArrayBufferViews = undefined,
+    outputs: MLNamedArrayBufferViews = undefined): tf.TensorContainerObject {
+    if (inputs) {
+      this.validateInputs(inputs);
+    } else {
+      inputs = {};
+      for (const inputName of this.inputs_.keys()) {
+        const inputOperand = this.inputs_.get(inputName);
+        const typedArrayConstructor =
+            utils.getTypedArray(inputOperand.desc.type);
+        const inputBuffer = new typedArrayConstructor(
+            utils.sizeFromDimensions(inputOperand.desc.dimensions));
+        inputs[inputName] = inputBuffer;
+      }
+    }
+    let outputOperands: Map<string, OutputOperand> = this.outputs_;
+    if (outputs) {
+      // Validate and filter the required output operands.
+      utils.assert(Object.keys(outputs).length !== 0,
+                   'The outputs is invalid.');
+      outputOperands = new Map();
+      for (const outputName in outputs) {
+        utils.assert(
+            typeof outputName === 'string' && this.outputs_.has(outputName),
+            'The name of the output is invalid.');
+        utils.assert(
+            utils.isTypedArray(outputs[outputName]),
+            'Only output of ArrayBufferView type is supported.');
+        outputOperands.set(outputName, this.outputs_.get(outputName));
+      }
     }
-
-    // Compute the output tensors.
     const outputTensors: tf.TensorContainerObject = tf.tidy(() => {
       const context = new ExecutionContext(
           this.constantTensors_, this.inputs_, inputs, this.operandRefs_);
       // The input and immediate tensors will be cleaned up.
       return context.compute(outputOperands);
     });
+    return outputTensors;
+  }
 
+  /** @internal */
+  async compute(
+    inputs: MLNamedArrayBufferViews,
+    outputs: MLNamedArrayBufferViews): Promise<void> {
+    const outputTensors: tf.TensorContainerObject =
+        this.computeOutputTensors(inputs, outputs);
     // Setup the outputs.
     for (const outputName of Object.keys(outputTensors)) {
       const tensor = outputTensors[outputName] as tf.Tensor;
       const desc = utils.createOperandDescriptorFromTensor(tensor);
-      const resource = outputs[outputName] as ArrayBufferView;
+      const data = await tensor.data();
+      const resource = outputs[outputName] ;
       utils.validateTypedArray(resource, desc.type, desc.dimensions);
-      resource.set(tensor.dataSync());
+      resource.set(data);
```

Comment:
Move Ln.181 closer to the data usage or just combine them into `resource.set(await tensor.data())`?

Reply:
Updated, thanks.